pkg/pkg/helm.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/pkg/errors"
+	"github.com/spf13/afero"
+	yaml "gopkg.in/yaml.v2"
+)
+
+type chartConfig struct {
+	Description string `json:"description"`
+}
+
+// Helm is a package based on a Helm chart.
+type Helm struct {
+	a            app.App
+	name         string
+	registryName string
+	config       chartConfig
+}
+
+var _ Package = (*Helm)(nil)
+
+// NewHelm creates an instance of Helm.
+func NewHelm(a app.App, name, registryName, version string) (*Helm, error) {
+	cp, err := chartConfigPath(a, name, registryName, version)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding chart path")
+	}
+
+	b, err := afero.ReadFile(a.Fs(), cp)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading chart configuration")
+	}
+
+	var cc chartConfig
+	if err = yaml.Unmarshal(b, &cc); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling chart configuration")
+	}
+
+	return &Helm{
+		a:            a,
+		name:         name,
+		registryName: registryName,
+		config:       cc,
+	}, nil
+}
+
+func chartConfigPath(a app.App, name, registryName, version string) (string, error) {
+	var err error
+	if version == "" {
+		version, err = latestChartRelease(a, name, registryName, version)
+		if err != nil {
+			return "", errors.Wrapf(err, "finding latest %s chart release", name)
+		}
+	}
+
+	chartConfigPath := filepath.Join(a.VendorPath(), registryName, name, "helm", version, name, "Chart.yaml")
+	return chartConfigPath, nil
+}
+
+func latestChartRelease(a app.App, name, registryName, version string) (string, error) {
+	chartPath := filepath.Join(a.VendorPath(), registryName, name, "helm")
+
+	fis, err := afero.ReadDir(a.Fs(), chartPath)
+	if err != nil {
+		return "", err
+	}
+
+	var versions []semver.Version
+	for _, fi := range fis {
+		if fi.IsDir() {
+			v, err := semver.Make(fi.Name())
+			if err != nil {
+				return "", err
+			}
+
+			versions = append(versions, v)
+		}
+	}
+
+	semver.Sort(versions)
+	return versions[0].String(), nil
+}
+
+// Name returns the name for the Helm chart.
+func (h *Helm) Name() string {
+	return h.name
+}
+
+// RegistryName returns the registry name for the Helm chart.
+func (h *Helm) RegistryName() string {
+	return h.registryName
+}
+
+// Description returns the description for the Helm chart. The description
+// is retrieved from the chart's Chart.yaml file.
+func (h *Helm) Description() string {
+	return h.config.Description
+}
+
+// IsInstalled returns true if the package is installed. a package is installed if it
+// has a libraries entry in app.yaml.
+func (h *Helm) IsInstalled() (bool, error) {
+	libs, err := h.a.Libraries()
+	if err != nil {
+		return false, errors.Wrapf(err, "checking if package %q is installed", h.Name())
+	}
+
+	_, isInstalled := libs[h.Name()]
+	return isInstalled, nil
+}
+
+func (h *Helm) prototypeName() string {
+	return fmt.Sprintf("io.ksonnet.pkg.%s-%s", h.registryName, h.name)
+}
+
+// Prototypes returns prototypes for this package. Currently, it returns a single prototype.
```

Comment:
Do we think that assumption will change for helm packages? (1:1)

Reply:
I'm not sure, but don't want to code myself in a corner.

Reply:
👍